### ISSUE TITLE
layer_norm_post_all_gather requires beta whenever gamma is provided for LayerNorm

### ISSUE THREAD

**Issue by sdjordjevicTT** [tenstorrent/tt-metal#41378]
### Description
ttnn.layer_norm_post_all_gather unconditionally rejects a gamma-without-beta call when norm_type=LAYERNORM, even though this is a mathematically valid configuration and the underlying kernel supports it.

In [layernorm_post_all_gather_device_operation.cpp:116](https://github.com/tenstorrent/tt-metal/blob/0e1451bb80a91a2bdbf4b45d3b024cc375deeb2f/ttnn/cpp/ttnn/operations/normalization/layernorm_distributed/device/layernorm_post_all_gather_device_operation.cpp#L116) (inside the if (gamma.has_value()) guard):

```cpp
const bool is_layernorm = args.norm_type == LayerNormDistributedType::LAYERNORM;
const bool has_beta     = beta.has_value();
TT_FATAL(is_layernorm == has_beta,
         "Beta tensor must be present if and only if using layernorm (vs rmsnorm)");
```
This forces beta to be present whenever gamma is provided and norm_type=LAYERNORM. The constraint conflates two independent concerns:

Both gamma and beta are optional in the standard LayerNorm formulation. A model may use LayerNorm with a learned scale but no bias, which is a common pattern (e.g. LLaMA-style norms).

The constraint is overly conservative. The kernel and program factory seems to already handle all four combinations independently:

- Kernel (layernorm_post_allgather.cpp): do_gamma and do_beta are separate compile-time args with distinct code paths for every combination (do_gamma && do_beta, do_gamma only, do_beta only, neither).
- Program factory (layernorm_post_all_gather_program_factory.cpp:316-317): passes gamma.has_value() and beta.has_value() independently.

The TT_FATAL on line 116 is the only place enforcing this artificial coupling.

### Repro
The following python test repros the issue:
```PY
# Reproducer for the bug where layer_norm_post_all_gather incorrectly requires
# beta when gamma is provided and norm_type is LAYERNORM.
#
# The overly strict constraint lives at:
#   ttnn/cpp/ttnn/operations/normalization/layernorm_distributed/device/
#   layernorm_post_all_gather_device_operation.cpp:116
#
#     TT_FATAL(is_layernorm == has_beta,
#              "Beta tensor must be present if and only if using layernorm (vs rmsnorm)");
#
# This fires even though the underlying kernel fully supports layernorm + gamma
# without beta (do_gamma and do_beta are independent compile-time flags).
#
# Pipeline under test (single device, n_devices=1):
#
#   input (1,1,37,72)
#       |
#   ttnn.layer_norm_pre_all_gather   → stats (1,1,64,64) [sum(x^2) | sum(x)]
#       |
#   [ttnn.all_gather would go here on real multi-device; skipped for 1 device]
#       |
#   ttnn.layer_norm_post_all_gather  ← gamma, (optional) beta
#       |
#   output (1,1,37,72)

import pytest
import torch
import ttnn


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------

def _make_stats_manually(x: torch.Tensor, n_devices: int = 1) -> torch.Tensor:
    """Build the stats tensor that pre_all_gather + all_gather would produce.

    For layernorm, each device contributes 2 tiles (64 cols):
      - tile 0 (cols  0.. 1): sum(x^2) in col 0
      - tile 1 (cols 32..33): sum(x)   in col 0
    """
    tile_cols_per_device = 2  # layernorm: sum(x^2) and sum(x)
    inp_shape = x.shape
    stats = torch.zeros(inp_shape[:-1] + (32 * n_devices * tile_cols_per_device,))
    chunk_w = inp_shape[-1] // n_devices
    for d in range(n_devices):
        chunk = x[..., d * chunk_w : (d + 1) * chunk_w]
        sumx2 = chunk.pow(2).sum(dim=-1, keepdim=True)  # (..., 1)
        sumx = chunk.sum(dim=-1, keepdim=True)           # (..., 1)
        base = d * tile_cols_per_device * 32
        stats[..., base : base + 1] = sumx2
        stats[..., base + 32 : base + 33] = sumx
    return stats


# ---------------------------------------------------------------------------
# test
# ---------------------------------------------------------------------------

@pytest.mark.parametrize("use_pre_all_gather", [True, False], ids=["via_pre_allgather", "stats_manual"])
def test_layernorm_post_allgather_no_beta_repro(use_pre_all_gather, device):
    """
    Demonstrates that layer_norm_post_all_gather incorrectly rejects
    gamma-without-beta for LayerNorm.

    Expected behaviour (once the bug is fixed):
      - gamma + beta  → OK
      - gamma only    → OK  (currently hits TT_FATAL)
      - neither       → OK

    All three cases are exercised.  The test passes only when case 2 no longer
    raises an exception.
    """
    torch.manual_seed(42)

    inp_shape = (1, 1, 37, 72)
    hidden_size = inp_shape[-1]   # 72
    epsilon = 1e-5

    dram = ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM)

    # ── input tensor ──────────────────────────────────────────────────────────
    x = torch.randn(inp_shape, dtype=torch.bfloat16)

    tt_inp = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device, memory_config=dram)

    # ── stats: either from the real pre_all_gather kernel or built manually ──
    if use_pre_all_gather:
        # On a single device the pre_all_gather output is directly usable by
        # post_all_gather (all_gather with n_devices=1 is a no-op).
        tt_stats = ttnn.layer_norm_pre_all_gather(tt_inp, dtype=ttnn.bfloat16, memory_config=dram)
    else:
        stats_torch = _make_stats_manually(x.float())
        tt_stats = ttnn.from_torch(
            stats_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device, memory_config=dram
        )

    # ── affine parameters (TILE layout; last dim padded 72→96 automatically) ─
    # gamma shape (1,1,1,72) → tiled padded shape (1,1,32,96)
    gamma_torch = torch.ones(1, 1, 1, hidden_size, dtype=torch.bfloat16)
    beta_torch = torch.zeros(1, 1, 1, hidden_size, dtype=torch.bfloat16)

    tt_gamma = ttnn.from_torch(gamma_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device, memory_config=dram)
    tt_beta = ttnn.from_torch(beta_torch, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device, memory_config=dram)

    common_kwargs = dict(epsilon=epsilon, memory_config=dram)

    # ── Case 1: gamma + beta (should pass before and after fix) ───────────────
    out1 = ttnn.layer_norm_post_all_gather(tt_inp, tt_stats, weight=tt_gamma, bias=tt_beta, **common_kwargs)
    assert out1 is not None, "Case 1 (gamma + beta) unexpectedly returned None"

    # ── Case 2: gamma only — THIS IS THE BUG ─────────────────────────────────
    # Currently hits TT_FATAL at:
    #   layernorm_post_all_gather_device_operation.cpp:116
    #   TT_FATAL(is_layernorm == has_beta, ...)
    # because is_layernorm=true but has_beta=false → false != true → fatal.
    #
    # The kernel itself supports this just fine (do_beta=0 is a valid path).
    out2 = ttnn.layer_norm_post_all_gather(tt_inp, tt_stats, weight=tt_gamma, **common_kwargs)
    assert out2 is not None, "Case 2 (gamma, no beta) unexpectedly returned None"

    # ── Case 3: no gamma, no beta (plain normalisation) ───────────────────────
    out3 = ttnn.layer_norm_post_all_gather(tt_inp, tt_stats, **common_kwargs)
    assert out3 is not None, "Case 3 (no gamma, no beta) unexpectedly returned None"
```

**Comment by sdjordjevicTT** [comment]
I checked this, @bbradelTT. It works. I am closing the issue. Thanks a lot!
